### PR TITLE
Foreground push notifications support and proper notification opening handling

### PR DESCRIPTION
**Pull request by jkadamczyk**
## What changed (plus any additional context for devs)
- Implemented showing push notifications in the foreground. Before they were getting lost when the application was in the foreground and not displayed at all.
- Implemented actions when the specific notification type is opened. In this case, when transaction notification is opened, the app will navigate to the transactions list and switch a wallet to the one it should given the notification data

## What to test

There are a couple of flows to check on both platforms:

### Wallet connect

- Connect Dapp:
	- App in foreground
	- Connect to a dapp in a web browser on your computer (QR Code)
	- Bottom sheet to connect should come up
- Sending TX in PC web browser:
	- App in foreground
	- Connect to a dapp on a web browser on your computer (QR Code)
	- Send any transaction from dapp
	- Confirmation sheet should come up
- Sending TX in PC web browser:
	- App in Background or App in Quit state
	- Try sending a tx in the connected dapp on your computer web browser
	- A notification should come up on your phone
	- Press on a notification
	- app should open to the tx confirmation sheet
- Sending TX in Mobile Web Browser:
	- App in Background
	- Navigate to a dapp on mobile web browser
	- Connect wallet via WC
	- App should open and sheet should be presented with connect confirmation sheet.
- Sending TX from Mobile Web Browser – quit state:
	- App in Background or App in Quit state
	- Navigate to a dapp on a mobile web browser
	- Try sending a tx
	- app should open with the tx confirmation sheet

### Transactions

- Receiving a notification when the app is in the foreground with a different wallet opened:
	- App in the foreground
	- Open a different wallet than the notification will have in the payload
	- Send yourself the notification in some way
	- click on the notification
	- wallet gets changed to the one related to the notification
	- App navigates to the transactions list screen
	- app opens the action sheet for that notification to go to etherscan etc.
- Receiving a notification when the app is in the foreground within the same wallet:
	- App in foreground
	- Be in the same wallet as the notification that will come in
	- Send yourself the notification
	- Click notification
	- Should navigate to the transactions list screen
	- Should open the action sheet with the transaction to navigate to etherscan etc.
- Receiving a notification when the app is the quit or background state:
	- App in Background or in quit state
	- Notification arrives
	- App gets opened
	- the wallet is switched to the one related to the notification
	- app navigates to the transactions list screen
	- action sheet is opened for that transaction to go to etherscan etc

## Final checklist

- [x] Assigned individual reviewers?
- [x] Added labels? (`team1/team2`, `critical path`, `release`, `dev QA`)
- [x] Did you test both iOS and Android?
- [x] If your changes are visual, did you check both the light and dark themes?
- [x] Added e2e tests? If not, please specify why – hard e2e notifications
- [ ] If you added new critical path files, did you update the CODEOWNERS file?
- [ ] If no `dev QA` label, did you add the PR to the QA Queue?

### CODE REVIEW

File: src/components/activity-list/ActivityList.js [hunk 1]
```diff
@@ -114,52 +114,58 @@ const ActivityList = ({
...
-    )
-  ) : (
-    <ActivityListEmptyState
-      emoji="👻"
-      label={lang.t('activity_list.empty_state.testnet_label')}
-    >
-      {header}
-    </ActivityListEmptyState>
-  );
+  if (network === networkTypes.mainnet || sections.length) {
```
Comment (jkadamczyk) [inline src/components/activity-list/ActivityList.js:117]: I replaced the nested ternary with the if else statement, seemed much more readable

File: ios/Rainbow/AppDelegate.mm [hunk 2]
```diff
@@ -133,7 +133,11 @@ - (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
 //Called when a notification is delivered to a foreground app.
 -(void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions options))completionHandler
 {
-  completionHandler(UNAuthorizationOptionSound | UNAuthorizationOptionAlert | UNAuthorizationOptionBadge);
+  if (@available(iOS 14.0, *)) {
```
Comment (jkadamczyk) [inline ios/Rainbow/AppDelegate.mm:136]: Fixes XCode warning about deprecated API usage

File: ios/Rainbow/AppDelegate.mm [hunk 3]
```diff
@@ -179,8 +183,20 @@ - (void)applicationDidBecomeActive:(UIApplication *)application{
   }
   // delete the badge
   [UIApplication sharedApplication].applicationIconBadgeNumber = 0;
-  // delete the notifications
-  [[UNUserNotificationCenter currentNotificationCenter] removeAllDeliveredNotifications];
+  // delete the notifications from WC
```
Comment (jkadamczyk) [inline ios/Rainbow/AppDelegate.mm:186]: Changes behavior, to clear only wallet connect notifications and leave all other notifications. Before we supported only wallet connect notifications and we were clearing all of the notifications

File: src/components/transaction-list/TransactionList.js [hunk 4]
```diff
@@ -137,110 +124,7 @@ export default function TransactionList({
     e => {
       const { index } = e.nativeEvent;
       const item = transactions[index];
-      const { hash, from, minedAt, network, pending, to, status, type } = item;
```
Comment (jkadamczyk) [inline src/components/transaction-list/TransactionList.js:140]: Moved it to a separate function

File: src/notifications/NotificationsHandler.tsx [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+  const notifeeForegroundEventListener = useRef<Callback>();
+
+  /*
+  We need to save some properties to a ref in order to have an up-to-date value
+  inside the event listener callbacks closure
+   */
+  syncedStateRef.current.wallets = wallets;
+  syncedStateRef.current.contacts = contacts;
+
+  useEffect(() => {
```
Comment (jkadamczyk) [inline src/notifications/NotificationsHandler.tsx:77]: Please review this carefully, this is all the logic for subscribing to all notifications

File: src/components/settings-menu/DevNotificationsSection.tsx [hunk 6]
```diff
@@ -78,8 +46,8 @@ const DevNotificationsSection = () => {
       [address]: { subscription: 'off', tx: state[address].tx },
     }));
 
-    firebaseUnsubscribeTopics('watcher', chainId, address);
-    firebaseUnsubscribeTopics('owner', chainId, address);
+    unsubscribeFromTransactionTopics('watcher', chainId, address);
+    unsubscribeFromTransactionTopics('owner', chainId, address);
```
Comment (estebanmino) [inline src/components/settings-menu/DevNotificationsSection.tsx:50]: @pugson maybe we should block these settings on your upcoming PR when user is on a testnet, we'll be taking the testnet chain id which doesn't make sense here

Comment (estebanmino) [reply src/components/settings-menu/DevNotificationsSection.tsx:50]: we'll also need to handle different chain ids when we support other chains

File: src/handlers/transactions.ts [hunk 7]
```diff
@@ -76,3 +98,126 @@ export const getTransactionMethodName = async (
...
+      blockExplorerAction,
+      ...(ios ? [TransactionActions.close] : []),
+    ];
+    if (showContactInfo) {
+      buttons.unshift(
+        contact
+          ? TransactionActions.viewContact
+          : TransactionActions.addToContacts
+      );
+    }
```
Comment (estebanmino) [inline src/handlers/transactions.ts:170]: i also dev QAed this, speed up and cancel is working as expected 👌

File: src/handlers/transactions.ts [hunk 8]
```diff
@@ -76,3 +98,126 @@ export const getTransactionMethodName = async (
...
+
+    showActionSheetWithOptions(
+      {
+        cancelButtonIndex: buttons.length - 1,
+        options: buttons,
+        title: pending
+          ? `${headerInfo.type}${
+              showContactInfo
+                ? ' ' + headerInfo.divider + ' ' + headerInfo.address
```
Comment (estebanmino) [inline src/handlers/transactions.ts:179]: wonder if it's more cleaner the way you did below like 

```

` ${headerInfo.divider} ${headerInfo.address}`
```

File: src/handlers/transactions.ts [hunk 9]
```diff
@@ -76,3 +98,126 @@ export const getTransactionMethodName = async (
...
+          case TransactionActions.cancel:
+            Navigation.handleAction(Routes.SPEED_UP_AND_CANCEL_SHEET, {
+              tx: transactionDetails,
+              type: 'cancel',
+            });
+            break;
+          case blockExplorerAction:
+            ethereumUtils.openTransactionInBlockExplorer(hash, network);
+            break;
+          default: {
```
Comment (estebanmino) [inline src/handlers/transactions.ts:216]: these cases are working as expected 👌

File: src/notifications/NotificationsHandler.tsx [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+  Event as NotifeeEvent,
+  EventType,
+} from '@notifee/react-native';
+import { getProviderForNetwork } from '@/handlers/web3';
+import { ethereumUtils, isLowerCaseMatch } from '@/utils';
+import { NewTransactionOrAddCashTransaction } from '@/entities/transactions/transaction';
+import {
+  TransactionDirection,
+  TransactionStatus,
+  TransactionType,
```
Comment (estebanmino) [inline src/notifications/NotificationsHandler.tsx:42]: `'TransactionType' is declared but its value is never read.ts`

File: src/notifications/NotificationsHandler.tsx [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+    }
+  };
+
+  const handleWalletConnectNotification = (
+    remoteMessage: FirebaseMessagingTypes.RemoteMessage
+  ) => {
+    const topic = remoteMessage?.data?.topic;
+
+    setTimeout(() => {
+      const requests = dispatch(requestsForTopic(topic));
```
Comment (estebanmino) [inline src/notifications/NotificationsHandler.tsx:141]: `Argument of type 'string | undefined' is not assignable to parameter of type 'string'.`

File: src/notifications/NotificationsHandler.tsx [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+
+      // casting data payload to type that was agreed on with backend
+      const data = (notification.data as unknown) as TransactionNotificationData;
+
+      const { wallets, contacts } = syncedStateRef.current;
+      const { accountAddress, nativeCurrency } = store.getState().settings;
+
+      let walletAddress: string | null = accountAddress;
+      if (!isLowerCaseMatch(accountAddress, data.address)) {
+        walletAddress = await wallets.switchToWalletWithAddress(data.address);
```
Comment (estebanmino) [inline src/notifications/NotificationsHandler.tsx:214]: `Type 'string | null | undefined' is not assignable to type 'string | null'.`

File: src/notifications/NotificationsHandler.tsx [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+    messaging().setBackgroundMessageHandler(onBackgroundRemoteNotification);
+    messaging().getInitialNotification().then(handleAppOpenedWithNotification);
+    notificationOpenedListener.current = messaging().onNotificationOpenedApp(
+      handleAppOpenedWithNotification
+    );
+    appStateListener.current = ApplicationState.addEventListener(
+      'change',
+      nextAppState => {
+        if (appState.current === 'background' && nextAppState === 'active') {
+          handleDeferredNotificationIfNeeded();
```
Comment (estebanmino) [inline src/notifications/NotificationsHandler.tsx:93]: 🥇

File: src/notifications/NotificationsHandler.tsx (hunk 11 above)
Comment (estebanmino) [reply src/notifications/NotificationsHandler.tsx:141]: just a type, WC working as expected

File: src/notifications/NotificationsHandler.tsx [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+        if (!isZero(status)) {
+          let direction = TransactionDirection.out;
+          if (
+            parsedTransaction?.from &&
+            parsedTransaction?.to &&
+            isLowerCaseMatch(parsedTransaction.from, parsedTransaction.to)
+          ) {
+            direction = TransactionDirection.self;
+          }
+          if (
```
Comment (estebanmino) [inline src/notifications/NotificationsHandler.tsx:289]: isn't this always true if `isLowerCaseMatch(parsedTransaction.from, parsedTransaction.to)` is also true? i feel like this needs an else if otherwise if `direction` is `TransactionDirection.self` it will be always overridden by `TransactionDirection.in`

File: src/notifications/NotificationsHandler.tsx [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+        resultTransaction.title = getTitle({
+          protocol: parsedTransaction.protocol,
+          status: resultTransaction.status,
+          type: parsedTransaction.type,
+        });
+        resultTransaction.pending = false;
+        resultTransaction.minedAt = minedAt;
+
+        if (resultTransaction) {
+          showTransactionDetailsSheet(
```
Comment (estebanmino) [inline src/notifications/NotificationsHandler.tsx:320]: 💯

File: src/notifications/NotificationsHandler.tsx [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+          gasPrice: rpcTransaction.gasPrice,
+          data: rpcTransaction.data,
+        };
+
+        const parsedTransaction = await parseNewTransaction(
+          newTransactionDetails,
+          nativeCurrency
+        );
+        const resultTransaction = { ...parsedTransaction };
+        const minedAt = Math.floor(Date.now() / 1000);
```
Comment (estebanmino) [inline src/notifications/NotificationsHandler.tsx:262]: i was thinking about a case where the notification is pressed lets' say 2 hours after the tx was confirmed, then this time wouldn't be accurate by 2 hours, but at that time zerion would probably get the tx with the right time. we could request the block timestamp to get the accurate time  but is probably ok to have it this way

File: src/notifications/NotificationsHandler.tsx (hunk 5 above)
Comment (estebanmino) [reply src/notifications/NotificationsHandler.tsx:77]: i went through methods one by one on dev mode and went over notifee docs to see if there was something obvious not being followed, all looking good 🥇

**Review by estebanmino — APPROVED** [review]
very nice 🔥  left some minor comments and dev qaed all these cases https://www.notion.so/rainbowdotme/Push-Notifications-opening-and-foreground-delivery-4d083777c6b3459d918a22feb1050b15#4a0f60a0ee1c47ca8e58b74ad4a2e5b4

File: src/components/settings-menu/DevNotificationsSection.tsx (hunk 6 above)
Comment (pugson) [reply src/components/settings-menu/DevNotificationsSection.tsx:50]: yeah that whole screen is getting removed in my PR and mainnet is being hardcoded as chain id for now

Comment (pugson) [reply src/components/settings-menu/DevNotificationsSection.tsx:50]: i can put in a small notice that says notifications are only working with mainnet if the user is on another chain

Comment (estebanmino) [reply src/components/settings-menu/DevNotificationsSection.tsx:50]: nice! i think just removing the entry point is enough, we've been completely removing other features on testnets

Comment (pugson) [reply src/components/settings-menu/DevNotificationsSection.tsx:50]: might be confusing ux when a menu disappears that was previously there. gonna ask design about that

**Review by pugson — APPROVED** [review]
tested with my personal wallet. seems to be working fine! didn't find anything more

**Comment by BrodyHughes** [comment]
Only issues found so far:

- iOS only: when app is in quit state some WC methods cause the app to open, but no WC modal to pop up (recording attached)
- When you delete a wallet it no longer appears in notifications debug, but you still get notifications for it.

https://www.loom.com/share/b661a592fccc45a7bfa6a0cb0858fa28

All test cases pass aside from these problems. Going to poke around a little more / try to find some edge cases before approving.

**Review by BrodyHughes — APPROVED** [review]
Looks good! Great work here all

**Review by osdnk — APPROVED** [review]
